cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/artifact/ArtifactInspector.java

```diff
@@ -293,74 +289,47 @@ private void inspectAdditionalPlugins(ArtifactId artifactId, Set<PluginClass> ad
-                while (resources.hasMoreElements()) {
-                  URL packageResource = resources.nextElement();
-
-                  // Only inspect classes in the top level jar file for Plugins.
-                  // The jar manifest may have packages in Export-Package that are loadable from the bundled jar files,
-                  // which is for classloading purpose. Those classes won't be inspected for plugin classes.
-                  // There should be exactly one of resource that match, because it maps to a directory on the FS.
-                  if (packageResource.getProtocol().equals("file")) {
-                    Iterator<String> classFiles = DirUtils.list(new File(packageResource.toURI()), "class").iterator();
-
-                    // Transform class file into class name and filter by @Plugin class only
-                    iterators.add(Iterators.filter(
-                      Iterators.transform(classFiles, new Function<String, String>() {
-                        @Override
-                        public String apply(String input) {
-                          return getClassName(currentPackage, input);
-                        }
-                      }), new Predicate<String>() {
-                        @Override
-                        public boolean apply(String className) {
-                          return isPlugin(className, pluginClassLoader);
-                        }
-                      }));
-                  }
-                }
-                if (!iterators.isEmpty()) {
-                  classIterator = Iterators.concat(iterators.iterator());
-                }
-              } catch (Exception e) {
-                // Cannot happen
-                throw Throwables.propagate(e);
-              }
-            }
-
-            try {
-              return pluginClassLoader.loadClass(classIterator.next());
-            } catch (ClassNotFoundException | NoClassDefFoundError e) {
-              // Cannot happen, since the class name is from the list of the class files under the classloader.
-              throw Throwables.propagate(e);
-            }
+  private Iterable<Class<?>> getPluginClasses(
+    final Collection<String> packages,
+    final PluginClassLoader pluginClassLoader) {
+    Predicate<String> nameCheckPredicate = getClassNameCheckPredicate(packages);
+    try (JarFile jarFile = new JarFile(pluginClassLoader.getTopLevelJar())) {
+      return jarFile
+        .stream()
+        .filter(entry -> !entry.isDirectory())
+        .map(ZipEntry::getName)
+        .filter(nameCheckPredicate)
+        .map(fileName -> StringUtils.removeEnd(fileName, ".class").replace('/', '.'))
+        .filter(className -> isPlugin(className, pluginClassLoader))
+        .map(className -> {
+          try {
+            return pluginClassLoader.loadClass(className);
+          } catch (ClassNotFoundException e) {
```

Comment:
We don't need to catch the NoClassDefFoundError like in the old code?

Reply:
Should not be needed as we catch Throwable one level higher, but I will add to ensure I don't miss something

Reply:
Done